{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import functools\n",
    "import json"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "def importing_json_files(file_wf):\n",
    "    f_wf = open(file_wf) #informations for nf\n",
    "    # returns JSON object as\n",
    "    # a dictionary\n",
    "    wf = json.load(f_wf)\n",
    "    f_wf.close\n",
    "    return wf\n",
    "\n",
    "#importing the process info\n",
    "proc_nf = importing_json_files('../json/nf_proc_shell_old.json')\n",
    "proc_nf_tool = importing_json_files('../json/nf_proc_tool_shell.json')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'name_process': 'MINIMAP2_ALIGN',\n",
       " 'string_process': 'process MINIMAP2_ALIGN {\\n    tag \"$meta.id\"\\n    label \\'process_medium\\'\\n\\n    conda (params.enable_conda ? \\'bioconda::minimap2=2.21 bioconda::samtools=1.12\\' : null)\\n    container \"${ workflow.containerEngine == \\'singularity\\' && !task.ext.singularity_pull_docker_container ?\\n        \\'https://depot.galaxyproject.org/singularity/mulled-v2-66534bcbb7031a148b13e2ad42583020b9cd25c4:1679e915ddb9d6b4abda91880c4b48857d471bd8-0\\' :\\n        \\'quay.io/biocontainers/mulled-v2-66534bcbb7031a148b13e2ad42583020b9cd25c4:1679e915ddb9d6b4abda91880c4b48857d471bd8-0\\' }\"\\n\\n    input:\\n    tuple val(meta), path(reads)\\n    path reference\\n    val bam_format\\n    val cigar_paf_format\\n    val cigar_bam\\n\\n    output:\\n    tuple val(meta), path(\"*.paf\"), optional: true, emit: paf\\n    tuple val(meta), path(\"*.bam\"), optional: true, emit: bam\\n    path \"versions.yml\"           , emit: versions\\n\\n    when:\\n    task.ext.when == null || task.ext.when\\n\\n    script:\\n    def args = task.ext.args ?: \\'\\'\\n    def prefix = task.ext.prefix ?: \"${meta.id}\"\\n    def input_reads = meta.single_end ? \"$reads\" : \"${reads[0]} ${reads[1]}\"\\n    def bam_output = bam_format ? \"-a | samtools sort | samtools view -@ ${task.cpus} -b -h -o ${prefix}.bam\" : \"-o ${prefix}.paf\"\\n    def cigar_paf = cigar_paf_format && !bam_format ? \"-c\" : \\'\\'\\n    def set_cigar_bam = cigar_bam && bam_format ? \"-L\" : \\'\\'\\n    \"\"\"\\n    minimap2 \\\\\\\\\\n        $args \\\\\\\\\\n        -t $task.cpus \\\\\\\\\\n        $reference \\\\\\\\\\n        $input_reads \\\\\\\\\\n        $cigar_paf \\\\\\\\\\n        $set_cigar_bam \\\\\\\\\\n        $bam_output\\n\\n\\n    cat <<-END_VERSIONS > versions.yml\\n    \"${task.process}\":\\n        minimap2: \\\\$(minimap2 --version 2>&1)\\n    END_VERSIONS\\n    \"\"\"\\n}',\n",
       " 'nb_lignes_process': 46,\n",
       " 'string_script': '    def args = task.ext.args ?: \\'\\'\\n    def prefix = task.ext.prefix ?: \"${meta.id}\"\\n    def input_reads = meta.single_end ? \"$reads\" : \"${reads[0]} ${reads[1]}\"\\n    def bam_output = bam_format ? \"-a | samtools sort | samtools view -@ ${task.cpus} -b -h -o ${prefix}.bam\" : \"-o ${prefix}.paf\"\\n    def cigar_paf = cigar_paf_format && !bam_format ? \"-c\" : \\'\\'\\n    def set_cigar_bam = cigar_bam && bam_format ? \"-L\" : \\'\\'\\n    \"\"\"\\n    minimap2 \\\\\\\\\\n        $args \\\\\\\\\\n        -t $task.cpus \\\\\\\\\\n        $reference \\\\\\\\\\n        $input_reads \\\\\\\\\\n        $cigar_paf \\\\\\\\\\n        $set_cigar_bam \\\\\\\\\\n        $bam_output\\n\\n\\n    cat <<-END_VERSIONS > versions.yml\\n    \"${task.process}\":\\n        minimap2: \\\\$(minimap2 --version 2>&1)\\n    END_VERSIONS\\n    \"\"\"',\n",
       " 'nb_lignes_script': 21,\n",
       " 'language_script': 'bash',\n",
       " 'tools': ['Minimap2'],\n",
       " 'tools_url': ['https://bio.tools/samtools', 'https://bio.tools/minimap2'],\n",
       " 'tools_dico': [{'name': 'SAMtools',\n",
       "   'uri': 'https://bio.tools/samtools',\n",
       "   'topic': [[{'uri': 'http://edamontology.org/topic_3168',\n",
       "      'term': 'Sequencing'},\n",
       "     {'uri': 'http://edamontology.org/topic_3325', 'term': 'Rare diseases'},\n",
       "     {'uri': 'http://edamontology.org/topic_0080',\n",
       "      'term': 'Sequence analysis'},\n",
       "     {'uri': 'http://edamontology.org/topic_0102', 'term': 'Mapping'}],\n",
       "    [{'uri': 'http://edamontology.org/topic_3168', 'term': 'DNA-Seq'},\n",
       "     {'uri': 'http://edamontology.org/topic_3325',\n",
       "      'term': 'https://en.wikipedia.org/wiki/Rare_disease'},\n",
       "     {'uri': 'http://edamontology.org/topic_0080', 'term': 'Sequences'}]],\n",
       "   'function': [{'operation': [[{'uri': 'http://edamontology.org/operation_3096',\n",
       "        'term': 'Editing'},\n",
       "       {'uri': 'http://edamontology.org/operation_0337',\n",
       "        'term': 'Visualisation'},\n",
       "       {'uri': 'http://edamontology.org/operation_1812', 'term': 'Parsing'},\n",
       "       {'uri': 'http://edamontology.org/operation_0335', 'term': 'Formatting'},\n",
       "       {'uri': 'http://edamontology.org/operation_0227', 'term': 'Indexing'}],\n",
       "      [{'uri': 'http://edamontology.org/operation_0337',\n",
       "        'term': 'Data visualisation'},\n",
       "       {'uri': 'http://edamontology.org/operation_0337', 'term': 'Rendering'},\n",
       "       {'uri': 'http://edamontology.org/operation_1812',\n",
       "        'term': 'Data loading'},\n",
       "       {'uri': 'http://edamontology.org/operation_1812', 'term': 'Loading'},\n",
       "       {'uri': 'http://edamontology.org/operation_0335',\n",
       "        'term': 'Format conversion'},\n",
       "       {'uri': 'http://edamontology.org/operation_0335',\n",
       "        'term': 'File reformatting'},\n",
       "       {'uri': 'http://edamontology.org/operation_0335',\n",
       "        'term': 'File format conversion'},\n",
       "       {'uri': 'http://edamontology.org/operation_0335',\n",
       "        'term': 'File formatting'},\n",
       "       {'uri': 'http://edamontology.org/operation_0335',\n",
       "        'term': 'Reformatting'},\n",
       "       {'uri': 'http://edamontology.org/operation_0227',\n",
       "        'term': 'Data indexing'},\n",
       "       {'uri': 'http://edamontology.org/operation_0227',\n",
       "        'term': 'Database indexing'}]],\n",
       "     'input': [{'uri': 'http://edamontology.org/data_0924',\n",
       "       'term': 'Sequence trace'}],\n",
       "     'output': [{'uri': 'http://edamontology.org/data_0924',\n",
       "       'term': 'Sequence trace'}]}],\n",
       "   'description': 'A software package with various utilities for processing alignments in the SAM format, including variant calling and alignment viewing.',\n",
       "   'homepage': 'http://www.htslib.org/'},\n",
       "  {'name': 'Minimap2',\n",
       "   'uri': 'https://bio.tools/minimap2',\n",
       "   'topic': [[{'uri': 'http://edamontology.org/topic_0102',\n",
       "      'term': 'Mapping'}],\n",
       "    []],\n",
       "   'function': [{'operation': [[{'uri': 'http://edamontology.org/operation_0491',\n",
       "        'term': 'Pairwise sequence alignment'}],\n",
       "      [{'uri': 'http://edamontology.org/operation_0491',\n",
       "        'term': 'Pairwise alignment'}]],\n",
       "     'input': [],\n",
       "     'output': []}],\n",
       "   'description': 'Pairwise aligner for genomic and spliced nucleotide sequences',\n",
       "   'homepage': 'https://github.com/lh3/minimap2'}],\n",
       " 'inputs': ['meta',\n",
       "  'reads',\n",
       "  'reference',\n",
       "  'bam_format',\n",
       "  'cigar_paf_format',\n",
       "  'cigar_bam'],\n",
       " 'nb_inputs': 6,\n",
       " 'outputs': [],\n",
       " 'nb_outputs': 0,\n",
       " 'name_workflow': 'ABMicroBioinf__magph',\n",
       " 'directive': ['tag \"$meta.id\"',\n",
       "  \"label 'process_medium'\",\n",
       "  \"conda (params.enable_conda ? 'bioconda::minimap2=2.21 bioconda::samtools=1.12' : null)\",\n",
       "  'container \"${ workflow.containerEngine == \\'singularity\\' && !task.ext.singularity_pull_docker_container ? \\'https://depot.galaxyproject.org/singularity/mulled-v2-66534bcbb7031a148b13e2ad42583020b9cd25c4:1679e915ddb9d6b4abda91880c4b48857d471bd8-0\\' : \\'quay.io/biocontainers/mulled-v2-66534bcbb7031a148b13e2ad42583020b9cd25c4:1679e915ddb9d6b4abda91880c4b48857d471bd8-0\\' }\"'],\n",
       " 'when': 'task.ext.when == null || task.ext.when',\n",
       " 'stub': '',\n",
       " 'code': 'process MINIMAP2_ALIGN {\\n    tag \"$meta.id\"\\n    label \\'process_medium\\'\\n\\n    conda (params.enable_conda ? \\'bioconda::minimap2=2.21 bioconda::samtools=1.12\\' : null)\\n    container \"${ workflow.containerEngine == \\'singularity\\' && !task.ext.singularity_pull_docker_container ?\\n        \\'https://depot.galaxyproject.org/singularity/mulled-v2-66534bcbb7031a148b13e2ad42583020b9cd25c4:1679e915ddb9d6b4abda91880c4b48857d471bd8-0\\' :\\n        \\'quay.io/biocontainers/mulled-v2-66534bcbb7031a148b13e2ad42583020b9cd25c4:1679e915ddb9d6b4abda91880c4b48857d471bd8-0\\' }\"\\n\\n    input:\\n    tuple val(meta), path(reads)\\n    path reference\\n    val bam_format\\n    val cigar_paf_format\\n    val cigar_bam\\n\\n    output:\\n    tuple val(meta), path(\"*.paf\"), optional: true, emit: paf\\n    tuple val(meta), path(\"*.bam\"), optional: true, emit: bam\\n    path \"versions.yml\"           , emit: versions\\n\\n    when:\\n    task.ext.when == null || task.ext.when\\n\\n    script:\\n    def args = task.ext.args ?: \\'\\'\\n    def prefix = task.ext.prefix ?: \"${meta.id}\"\\n    def input_reads = meta.single_end ? \"$reads\" : \"${reads[0]} ${reads[1]}\"\\n    def bam_output = bam_format ? \"-a | samtools sort | samtools view -@ ${task.cpus} -b -h -o ${prefix}.bam\" : \"-o ${prefix}.paf\"\\n    def cigar_paf = cigar_paf_format && !bam_format ? \"-c\" : \\'\\'\\n    def set_cigar_bam = cigar_bam && bam_format ? \"-L\" : \\'\\'\\n    \"\"\"\\n    minimap2 \\\\\\\\\\n        $args \\\\\\\\\\n        -t $task.cpus \\\\\\\\\\n        $reference \\\\\\\\\\n        $input_reads \\\\\\\\\\n        $cigar_paf \\\\\\\\\\n        $set_cigar_bam \\\\\\\\\\n        $bam_output\\n\\n\\n    cat <<-END_VERSIONS > versions.yml\\n    \"${task.process}\":\\n        minimap2: \\\\$(minimap2 --version 2>&1)\\n    END_VERSIONS\\n    \"\"\"\\n}',\n",
       " 'shell': '\\n    minimap2 \\\\\\\\\\n        -t \\\\\\\\\\n    cat <<-END_VERSIONS \\n versions.yml\\n        minimap2: \\\\$(minimap2 --version 2\\n&1)\\n    END_VERSIONS\\n',\n",
       " 'wf_orig': 'magph',\n",
       " 'name': 'ABMicroBioinf__magph/MINIMAP2_ALIGN',\n",
       " 'file': '/home/marinedjaffardjy/Documents/AnalyseDonneesNextflow/results_nf/new_analysis_nf/ABMicroBioinf__magph/processes_info.json',\n",
       " 'owner': 'ABMicroBioinf',\n",
       " 'shell_modif': 'minimap2\\\\\\\\-t\\\\\\\\cat<<-END_VERSIONSversions.ymlminimap2:\\\\$(minimap2--version2&1)END_VERSIONS'}"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "proc_nf_tool[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "#take out the tools not in the dicts\n",
    "for proc in proc_nf_tool:\n",
    "    tools = proc[\"tools\"]\n",
    "    tools_dict = []\n",
    "    for el in proc[\"tools_dico\"]:\n",
    "        if el[\"name\"] in tools:\n",
    "            tools_dict.append(el)\n",
    "    proc[\"tools_dico\"] = tools_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'uri': 'http://edamontology.org/topic_0102', 'term': 'Mapping'}]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "proc_nf_tool[0][\"tools_dico\"][0][\"topic\"][0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "topics_repart_nf = []\n",
    "for proc in proc_nf_tool :\n",
    "    topics = []\n",
    "    for el in proc['tools_dico'] :\n",
    "        for top in el[\"function\"][0][\"operation\"]:\n",
    "            if(len(top)>0):\n",
    "                topics.append(top[0][\"term\"])\n",
    "    topics=list(set(topics))\n",
    "    topics_repart_nf.append(topics)\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_topics=[]\n",
    "for el in topics_repart_nf:\n",
    "    for top in el:\n",
    "        all_topics.append(top)\n",
    "all_topics = list(set(all_topics))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [],
   "source": [
    "nb_use = []\n",
    "for topic in all_topics:\n",
    "    count = 0\n",
    "    for liste in topics_repart_nf:\n",
    "        if topic in liste:\n",
    "            count+=1\n",
    "    nb_use.append({\"operation\": topic, \"count\":count})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_repart_topics=pd.DataFrame(nb_use)\n",
    "df_repart_topics = df_repart_topics.sort_values(by=\"count\", ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>operation</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>197</th>\n",
       "      <td>Data visualisation</td>\n",
       "      <td>1948</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>149</th>\n",
       "      <td>Editing</td>\n",
       "      <td>1935</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>225</th>\n",
       "      <td>Sequence analysis (general)</td>\n",
       "      <td>1014</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>93</th>\n",
       "      <td>Polymorphism detection</td>\n",
       "      <td>978</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>182</th>\n",
       "      <td>Significance testing</td>\n",
       "      <td>816</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>Statistical calculation</td>\n",
       "      <td>816</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>232</th>\n",
       "      <td>Validation</td>\n",
       "      <td>731</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>172</th>\n",
       "      <td>Utility operation</td>\n",
       "      <td>720</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85</th>\n",
       "      <td>Data handling</td>\n",
       "      <td>653</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>194</th>\n",
       "      <td>Sequence alignment generation</td>\n",
       "      <td>529</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>Sequence alignment</td>\n",
       "      <td>524</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>Genome indexing</td>\n",
       "      <td>361</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>126</th>\n",
       "      <td>Construction</td>\n",
       "      <td>358</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>187</th>\n",
       "      <td>Cartography</td>\n",
       "      <td>296</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>263</th>\n",
       "      <td>Mapping</td>\n",
       "      <td>295</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>282</th>\n",
       "      <td>Sequence variation analysis</td>\n",
       "      <td>272</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>295</th>\n",
       "      <td>Genetic variation analysis</td>\n",
       "      <td>270</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>Expression data analysis</td>\n",
       "      <td>182</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69</th>\n",
       "      <td>Expression analysis</td>\n",
       "      <td>169</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>Trimming</td>\n",
       "      <td>168</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                         operation  count\n",
       "197             Data visualisation   1948\n",
       "149                        Editing   1935\n",
       "225    Sequence analysis (general)   1014\n",
       "93          Polymorphism detection    978\n",
       "182           Significance testing    816\n",
       "49         Statistical calculation    816\n",
       "232                     Validation    731\n",
       "172              Utility operation    720\n",
       "85                   Data handling    653\n",
       "194  Sequence alignment generation    529\n",
       "58              Sequence alignment    524\n",
       "36                 Genome indexing    361\n",
       "126                   Construction    358\n",
       "187                    Cartography    296\n",
       "263                        Mapping    295\n",
       "282    Sequence variation analysis    272\n",
       "295     Genetic variation analysis    270\n",
       "55        Expression data analysis    182\n",
       "69             Expression analysis    169\n",
       "45                        Trimming    168"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_repart_topics.head(20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
